lib/pacemaker/pcmk_rule.c

```diff
@@ -0,0 +1,290 @@
+
+done:
+    free(xpath);
+    freeXpathObject(xpath_obj);
+    return rc;
+}
+
+/*!
+ * \internal
+ * \brief Initialize the cluster working set for checking rules
+ *
+ * Make our own copies of the CIB XML and date/time object, if they're not
+ * \c NULL. This way we don't have to take ownership of the objects passed via
+ * the API.
+ *
+ * \param[in,out] out       Output object
+ * \param[in]     input     The CIB XML to check (if \c NULL, use current CIB)
+ * \param[in]     date      Check whether the rule is in effect at this date
+ *                          and time (if \c NULL, use current date and time)
+ * \param[out]    data_set  Where to store the cluster working set
+ *
+ * \return Standard Pacemaker return code
+ */
+static int
+rule_init(pcmk__output_t *out, xmlNodePtr input, crm_time_t *date,
+          pe_working_set_t **data_set)
+{
+    // Allows for cleaner syntax than dereferencing the data_set argument
+    pe_working_set_t *new_data_set = NULL;
+
+    CRM_ASSERT((out != NULL) && (data_set != NULL));
+
+    new_data_set = pe_new_working_set();
+    if (new_data_set == NULL) {
+        return ENOMEM;
+    }
+
+    pe__set_working_set_flags(new_data_set,
+                              pe_flag_no_counts|pe_flag_no_compat);
+
+    // Populate the working set instance
+
+    // Make our own copy of the given input or fetch the CIB and use that
+    if (input != NULL) {
+        new_data_set->input = copy_xml(input);
+        if (new_data_set->input == NULL) {
+            out->err(out, "Failed to copy input XML");
+            pe_free_working_set(new_data_set);
+            return ENOMEM;
+        }
+
+    } else {
+        int rc = cib__signon_query(NULL, &(new_data_set->input));
+
+        if (rc != pcmk_rc_ok) {
+            out->err(out, "CIB query failed: %s", pcmk_rc_str(rc));
+            pe_free_working_set(new_data_set);
+            return rc;
+        }
+    }
+
+    // Make our own copy of the given crm_time_t object or use the current time
+    if (date != NULL) {
+        // pcmk_copy_time() guarantees non-NULL
+        new_data_set->now = pcmk_copy_time(date);
+
+    } else {
+        // So does crm_time_new()
+        new_data_set->now = crm_time_new(NULL);
+    }
+
+    // Unpack everything
+    cluster_status(new_data_set);
+
+    *data_set = new_data_set;
+
+    return pcmk_rc_ok;
+}
+
+/*!
+ * \internal
+ * \brief Check whether each rule in a list is in effect
+ *
+ * \param[in,out] out       Output object
+ * \param[in]     input     The CIB XML to check (if \c NULL, use current CIB)
+ * \param[in]     date      Check whether the rule is in effect at this date and
+ *                          time (if \c NULL, use current date and time)
+ * \param[in]     rule_ids  The IDs of the rules to check
+ *
+ * \return Standard Pacemaker return code
+ */
+int
+pcmk__check_rules(pcmk__output_t *out, xmlNodePtr input, crm_time_t *date,
+                  const char **rule_ids)
```

Comment:
I think it makes more sense for the API functions should handle just one rule at a time, instead of processing the whole list at once as `crm_rule` does. That way, we get to keep the return code for every rule we check, instead of just "success or the last non-success return code."

For this draft, I wrote it to accept a list of rule IDs like the existing logic in `crm_rule`. That way the output structure doesn't change. This can easily be updated to take just one rule ID though.

Reply:
Updated to add static wrappers for calling with a single rule ID.